keps/sig-auth/2579-psp-replacement/README.md

```diff
@@ -0,0 +1,969 @@
+
+### PodSecurityPolicy Migration
+
+Migrating to the replacement policy from PodSecurityPolicies can be done effectively using a
+combination of dry-run and audit/warn modes (although this becomes harder if mutating PSPs are
+used).
+
+We could also ship a standalone tool to assist with the PodSecurityPolicy migration. Here are some
+ideas for the sorts of things the tool could assist with:
+
+- Analyze PSP resources, identify the closest profile level, and highlight the differences
+- Check the authorization mode for existing pods. For example, if a pod’s service account is not
+  authorized to use the PSP that validated it (based on the `kubernetes.io/psp` annotation), then
+  that should trigger a warning.
+- Automate the dry-run and/or labeling process.
+- Automatically select (and optionally apply) a policy level for each namespace.
+
+We should also publish a step-by-step migration guide. A rough approach might look something like
+this, with the items tagged (automated) having support from the PSP migration tool.
+
+1. Enable 3-tier policy admission plugin, default everything to privileged.
+2. Eliminate mutating PSPs:
+    1. Clone all mutating PSPs to a non-mutating version (automated)
+    2. Update all ClusterRoles authorizing use of the mutating PSPs to also authorize use of the
+       non-mutating variant (automated)
+    3. Watch for pods using the mutating PSPs (check via the `kubernetes.io/psp` annotation), and
+       work with code owners to migrate to valid non-mutating resources.
+    4. Delete mutating PSPs
+3. Select a compatible profile for each namespace, based on the existing resources in the namespace
+   (automated)
+    1. Review the profile choices
+    2. Evaluate the difference in privileges that would come from disabling the PSP controller
+       (automated).
+4. (optional) Apply the profiles in `warn` and `audit` mode (automated)
+5. Apply the profiles in `allow` mode (automated)
+6. Disable PodSecurityPolicy
+
+### Custom Profiles
+
+Allow custom profile levels to be statically configured. E.g.
+`--extra-pod-security-levels=host-network`. Custom profiles are ignored by the built-in admission
+plugin, and must be handled completely by a 3rd party webhook (including the dry-run implementation,
+if desired).
+
+### Custom Warning Messages
+
+An optional `prefix/warn-message` annotation can be used to return a custom warning message (in
+addition to the standard message) whenever a policy warning is triggered. This could be useful for
+announcing the date that new policy will take effect, or providing a point of contact if you need to
+request an exception.
+
+### Windows Support
+
+Even without built-in support, enforcement for Windows pods can be delegated to a webhook admission
+plugin by exempting the `windows` RuntimeClass. We should investigate built-in Windows support out
+of the box though. Requirements include:
+
+1. A standardized identifier for Windows pods
```

Comment:
A `kubernetes.io/os = windows` node selector is the recommended way to identify a Windows pod.
https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host

I did see above that one of the goals is to `Don’t automatically break windows pods` 
For pod security policies that could potentially break Windows pods (for example restricting runAsUser ranges for all pods) it might be nice to skip enforcement of these if there is a `kubernetes.io=windows` nodeSelector.

Reply:
> A `kubernetes.io/os = windows` node selector is the recommended way to identify a Windows pod.

There's also https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#simplifying-with-runtimeclass  which recommends setting that selector via a RuntimeClass.

Reply:
We had a good discussion on this topic over slack [1], and concluded that we cannot safely rely on nodeselectors to bypass the policy since nodeSelectors themselves can easily be bypassed.

[1] https://kubernetes.slack.com/archives/C0SJ4AFB7/p1617830466145800

Reply:
> We had a good discussion on this topic over slack [1], and concluded that we cannot safely rely on nodeselectors to bypass the policy since nodeSelectors themselves can easily be bypassed.

In that conversation we identified a solution that could prevent nodeSelectors from being bypassed.

If we were to create an built-in admission controller that would restrict who could admit pods with `nodeName` set that could prevent nodeselectors from being bypassed.

> There's also https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#simplifying-with-runtimeclass which recommends setting that selector via a RuntimeClass.

This documentation shows how runtime classes can be used on Windows but it doesn't necessarily recommend using them for all clusters.
Windows containers must be built for a specific Windows OS version must be run on a matching OS version.  
Some clusters may contain nodes running multiple Windows OS versions and in that case a nodeSelector for `kubernetes.io/os=windows` is not sufficient for ensuring Windows pods/containers will land on compatible nodes. 
For these clusters it may be easier to ensure pods land on the correct nodes with runtime classes (versus ensuring all pods have a `node.kubernetes.io/windows-built` nodeSelector.

For the a general purpose way to identify Windows pods I still feel that using a `kubernets.io/os=windows` nodeSelector is the better option.

`kubernetes.io/os` labels are applied to nodes automatically so it is safe to recommend that all pod specs targeting Windows nodes always include a `kubernetes.io/os=windows` nodeSelector. 

If a built-in admission controller that restricts who can bypass the scheduled / submitting pods that have nodeName already filled and restricting who can edit node labels this should be a safe way to identify windows Pods for policy enforcement/exception.

Runtime classes must be manually added to a cluster so we cannot recommend that all window pod-specs include a runtimeClassName. Additionally (except for with dockershim which is being deprecated) runtime class handlers must be configured in the CRI's configuration files. 

One big concern I have for having policy exception be based on runtime classes is the scenario where someone has a cluster with Windows nodes/pods but is not using runtime classes - and then would like to enable policy.
If they just enable policy here their pods would not be excepted from policy and would fail.
Users would then need to update all the Windows pod specs to add runtimeClassNames.
If that cluster is running containerd and runtime class handlers are not defined in the config.toml configuration file then they would need to reconfigure/redeploy all of their Windows nodes as well.
Just enabling policy on an existing cluster could be very painful for operations.

Additionally I could see other benefits to having the above mentioned built-in admission controller.

Reply:
> If we were to create an built-in admission controller that would restrict who could admit pods with `nodeName` set that could prevent nodeselectors from being bypassed.

however, that could not be enabled by default for backwards compatibility reasons, which would mean this mechanism could not rely on it by default

Reply:
> If they just enable policy here their pods would not be excepted from policy and would fail.

Ft's important to note that enabling the mechanism proposed in this KEP doesn't fail anything by default... you have to opt into enforcement levels per-namespace for it to affect existing workloads. Having a direct mechanism to identify pods that should be exempted (e.g. runtime class, not node selector) makes sense to me, and can be rolled out gradually for existing windows clusters

Reply:
> > If we were to create an built-in admission controller that would restrict who could admit pods with `nodeName` set that could prevent nodeselectors from being bypassed.
> 
> however, that could not be enabled by default for backwards compatibility reasons, which would mean this mechanism could not rely on it by default

Even if this is not enabled by default - enabling it in an existing cluster would be a lot less disruptive than editing needing to potentially edit many pod-specs and reconfiguring nodes.

Reply:
> If that cluster is running containerd and runtime class handlers are not defined in the config.toml configuration file then they would need to reconfigure/redeploy all of their Windows nodes as well.

I don't think this is the case, because every containerd deployment will have atleast one default runtime (by [default](https://github.com/containerd/containerd/blob/88880f0f2cf7212d8f188083cb24dd41ad33553a/pkg/cri/config/config_windows.go#L40) it is `runhcs-wcow-process`

> Having a direct mechanism to identify pods that should be exempted (e.g. runtime class, not node selector) makes sense to me,

Why runtime class over node-selector (with correct enforcement)?  Runtime classes would require additional configuration and are not standardized across providers.  The node label is standardized on `kubernetes.io/os: 'windows'`.   The node selector administration controller wouldn't be backwards compatible but isn't PSP something that will be opt in as well?  I was surprised to find out that node selectors weren't enforced, I wonder if others would be too and could be an overall improvement in that case?

In both scenarios there would be additional work required:  

- For runtime-classes, if you are not already using them you would need to create one and update all your pod specs. 
- For node-selectors, you would need to turn on a new (possibility breaking) administration controller.  

IMO Both are disruptive and all that not user friendly.  I am going to bring this up at sig-windows this week to get feedback from the community.

side note: It seems like we run into these issues when identifying pods when ever discussing differences between linux/windows,  Is it possible we should revisit having OS field in the pod spec?

Reply:
We're already planning to bake in RuntimeClass based exceptions, windows pods aside. I definitely do not want to add generic nodeSelector based exceptions. I do agree that forbidding setting nodeName on create (or at least requiring bind permission) would be a reasonable thing to add, but I'm still not sure that addresses all my concerns with using labels.

IMO this thread is really highlighting the need for something along the lines of a `WindowsSecurityContext` or windows-specific pod options that could be used to definitively say whether a pod was a windows pod.

I would be more open keying off a label selector if we also added a kubelet admission check that prevented a non-windows kubelet from executing a pod with a windows node selector (or vice-versa)

Reply:
If we went the Runtimeclass path, it would require Adding a Runtimeclass to every pod (even if a pod had a windows node selector) otherwise they would be evaluated against the Linux Policies and possibly fail. Is that correct?  

What happens if they don't?  This came up in gatekeeper implementation as well (https://github.com/open-policy-agent/gatekeeper-library/pull/70#issuecomment-824460250). The solution there was to apply a policy that forced all pods to have some type of OS identifier.

> IMO this thread is really highlighting the need for something along the lines of a WindowsSecurityContext or windows-specific pod options that could be used to definitively say whether a pod was a windows pod.

agreed.  I assume that should be a separate KEP, and wouldn't be blocking here.

Reply:
Another open issue brought up at yesterday meeting was the scenario where a cluster has runtime handlers with the same name across different types of nodes.  This is an issue in a Dockershim deployment, since the runtimehandler is the same across Linux and Windows.  A Linux pod could by pass the policies by using a Windows runtime class (which is exempt) and setting a `nodename` selector.  

This appears to be an issue in not only mixed Windows/Linux clusters but also a Linux only cluster that was using Runtime classes as an exemption and has multiple runtimehandlers defined on a node.

Reply:
I agree with @marosset here that enabling this feature in clusters with  Windows pods which are not having runtime classes in problematic.

> I would be more open keying off a label selector if we also added a kubelet admission check that prevented a non-windows kubelet from executing a pod with a windows node selector (or vice-versa)

I think this is reasonable, we already strip linux specific security constraints if the kubelet is running on Windows nodes. So, it won't be too hard to implement. +1 to this suggestion. Anyone here has any objections to this suggestion?

Reply:
> enabling this feature in clusters with Windows pods which are not having runtime classes is problematic

Could you expand on why?

Reply:
Because the new admission controller can accidentally add a Linux Option that can break a Windows pod because admission controller cannot distinguish a linux pod from Windows pod. We were also thinking of having a default runtime class for linux etc but having field in pod Spec would be ideal. I started this discussion sometime ago on [slack](https://kubernetes.slack.com/archives/C0SJ4AFB7/p1605100181326900)

We had a similar problem in OpenShift, we worked around it by stripping those Linux Security Options added by admission controller in Kubelet if it is running on Windows node.

Reply:
> Because the new admission controller can accidentally add a Linux Option

this admission plug-in is non-mutating, so that would not happen.

Reply:
Thank you for the clarification. However isn't the problem here `Identifying if the pod is a Windows pod or Linux pod`. I believe the author wants to exclude Windows pods from admission plugin and the question is how to do it.

Reply:
Some problems I foresee, and questions I'm considering:
- imagine if Windows nodes became able to run container images that were made for Linux (this doesn't sound totally infeasible). If we, say, looked for any Pod with scheduling constraints to only run on nodes which run Windows, and made those Pods exempt from PodSecurityPolicy, then the Linux-compatible pods scheduled to such nodes could be able to bypass a security control. For example, they might be able to run as `root` even when the cluster admin expected to constrain them to the _Restricted_ pod security standard.
- if we have a future Windows-specific pod security standard, how do we apply it if Pods targeting `kubernetes.io/os=windows` nodes are exempt?
- it might be or become possible to make a dual-platform container image (eg using something like https://github.com/jart/cosmopolitan) and want it to run on whatever kind of node is available?\
  Imagine if that becomes common, perhaps with WASM / WASI executables. Does the PSP replacement policy design easily allow for this future or make it more difficult?

Reply:
> imagine if Windows nodes became able to run container images that were made for Linux (this doesn't sound totally infeasible).

Running Linux containers on Windows is actually possible today but not in kubernetes. There are other challenges in enabled this scenario in kubernetes that I won't go into here but the points @sftim raised are valid.